File: src/main/java/com/github/iamniklas/liocore/procedures/Procedure.java
package com.github.iamniklas.liocore.procedures;

import com.github.iamniklas.liocore.led.LEDDataBundle;

public abstract class Procedure {
    //TODO
    //LEDStripManager strip;
    LEDDataBundle dataBundle;
    ProcedureCalls procCalls;
    int step = 0;
    int steps = 0;
    int modulo = 1;
    boolean moduloInvert = false;
    boolean isSubProcedure = false;

    public Procedure(LEDDataBundle _bundle) {
        dataBundle = _bundle;

        //TODO Init. Strip
        procCalls = _bundle.procedureCalls;
        if(_bundle.puModulo != null) {
            modulo = _bundle.puModulo;
        }
        if(_bundle.puModuloInvert != null) {
            moduloInvert = _bundle.puModuloInvert;
        }
    }

    public abstract void start();
    public abstract void update();

    public void postUpdate() {
        if (modulo < 2) return;
        //TODO Get LED Strip size from LEDManager
        for (int i = 0; i < 300; i++) {
            if (moduloInvert) {
                if (i % modulo != 0) {
                    //mStrip!!.setPixel(i, com.github.iamniklas.liocorekotlin.led.ColorRGB.black.toSystemColor())
                }
            } else {
                if (i % modulo == 0) {
                    //mStrip!!.setPixel(i, com.github.iamniklas.liocorekotlin.led.ColorRGB.black.toSystemColor())
                }
            }
        }
        if (step > steps) {
            finishProcedure();
        }
    }

    protected void finishProcedure(boolean _clearStrip) {
        //TODO Impl. Strip Interface
        //mStrip!!.mProcContainer.removeCurrentProcedure()
        if (_clearStrip) {
            //mStrip!!.setAllPixels(com.github.iamniklas.liocorekotlin.led.ColorRGB.black.toSystemColor())
        }
        procCalls.onProcedureFinish();
    }

    protected void finishProcedure() {
        //TODO Impl. Strip Interface
        //mStrip!!.mProcContainer.removeCurrentProcedure()
        procCalls.onProcedureFinish();
    }
}

File: src/main/java/com/github/iamniklas/liocorekotlin/procedures/Procedure.kt
package com.github.iamniklas.liocorekotlin.procedures

import com.github.iamniklas.liocorekotlin.led.ColorRGB
import com.github.iamniklas.liocorekotlin.led.LEDDataBundle
import com.github.iamniklas.liocorekotlin.led.LEDStripManager

abstract class Procedure(_bundle: com.github.iamniklas.liocorekotlin.led.LEDDataBundle) {
    var mStrip: com.github.iamniklas.liocorekotlin.led.LEDStripManager?
    var mCallbacks: com.github.iamniklas.liocorekotlin.procedures.ProcedureCalls?
    var mStep = 0
    open var mSteps = 0
    open var mModulo = 1
    var mModuloInvert = false
    protected var mIsSubProcedure = false
    abstract fun start()
    abstract fun update()
    fun postUpdate() {
        if (mModulo < 2) return
        for (i in 0 .. com.github.iamniklas.liocorekotlin.led.LEDStripManager.LED_COUNT) {
            if (mModuloInvert) {
                if (i % mModulo != 0) {
                    mStrip!!.setPixel(i, com.github.iamniklas.liocorekotlin.led.ColorRGB.black.toSystemColor())
                }
            } else {
                if (i % mModulo == 0) {
                    mStrip!!.setPixel(i, com.github.iamniklas.liocorekotlin.led.ColorRGB.black.toSystemColor())
                }
            }
        }
        if (mStep > mSteps) {
            finishProcedure()
        }
    }

    protected fun finishProcedure(_clearStrip: Boolean) {
        mStrip!!.mProcContainer.removeCurrentProcedure()
        if (_clearStrip) {
            mStrip!!.setAllPixels(com.github.iamniklas.liocorekotlin.led.ColorRGB.black.toSystemColor())
        }
        mCallbacks!!.onProcedureFinish()
    }

    protected fun finishProcedure() {
        mStrip!!.mProcContainer.removeCurrentProcedure()
        mCallbacks!!.onProcedureFinish()
    }

    init {
        mStrip = _bundle.ledStrip
        mCallbacks = _bundle.procedureCalls
        if (_bundle.puModulo != null) {
            mModulo = _bundle.puModulo!!
        }
        if (_bundle.puModuloInvert != null) {
            mModuloInvert = _bundle.puModuloInvert!!
        }
    }
}